src/package-linker.js

```diff
@@ -91,18 +98,68 @@ export default class PackageLinker {
       }
     }
 
-    // no deps to link
-    if (!deps.length) {
-      return;
+    return deps;
+  }
+
+  async linkDependencies(flatTree: HoistManifestTuples): Promise<void> {
+    // Create a map of .bin location to a map of bin command and its
+    // source location. we remove all duplicates for each .bin location,
+    // this will prevents multiple calls to create the same symlink.
+    const binsByLocation: Map<string, Map<string, Manifest>> = new Map();
+    async function getBinLinks(binLoc: string): Promise<Map<string, Manifest>> {
+      if (!binsByLocation.has(binLoc)) {
+        // ensure our .bin file we're writing these to exists
+        await fs.mkdirp(binLoc);
+        binsByLocation.set(binLoc, new Map());
+      }
+      const binLinks = binsByLocation.get(binLoc);
+      invariant(binLinks, 'expected value');
+      return binLinks;
     }
 
-    // ensure our .bin file we're writing these to exists
-    const binLoc = path.join(dir, '.bin');
-    await fs.mkdirp(binLoc);
+    let binsCount = 0;
+    for (const [dest, {pkg}] of flatTree) {
+      const modules = this.config.getFolder(pkg);
+      const rootLoc = path.join(this.config.cwd, modules);
+      const pkgLoc = path.dirname(dest);
+      const parentBinLoc = path.join(pkgLoc, '.bin');
+      const pkgBinLoc = path.join(dest, modules, '.bin');
+
+      const deps = await this.getBinDependencies(pkg);
+
+      for (const {dep, loc} of deps) {
+        // replace dependency location that point to different package,
+        // if the one in the current location is identical
+        let newDepLoc;
+        const depLoc = path.dirname(loc);
+        if (depLoc !== pkgLoc && depLoc !== rootLoc && path.dirname(depLoc) !== dest) {
+          newDepLoc = path.join(dest, modules, dep.name);
+          const manifest = await this.config.maybeReadManifest(newDepLoc);
+          if (!manifest || manifest.version !== dep.version) {
+            newDepLoc = null;
+          }
+        }
+
+        // When both package and dependency are in the same folder use the .bin
+        // in that folder, else use the .bin in the package.
+        const location = newDepLoc || loc;
+        const binLoc = path.dirname(location) === pkgLoc ? parentBinLoc : pkgBinLoc;
+        const binLinks = await getBinLinks(binLoc);
+        // Remove Duplicates
+        if (!binLinks.has(location)) {
```

Comment:
If that works then sure.
I thought it would be easier though if you iterate over the top dependencies first to generate a set of bins and then go through deeper dependencies discarding bins that are already in the set.